### DISCUSSION TITLE
Libraries for inline_tests only

### DISCUSSION THREAD

**Discussion by wokalski** [ocaml/dune#4829] [Ideas]
I have a couple of modules which contain resources for tests. I don't want to bloat the library/executable so I don't want them included in the library itself when building for release. I currently put those files, along with the tests that use them, in a separate library. If I understand correctly this causes longer builds - a change in the main library itself, causes all tests to be recompiled/reran.

If I understand correctly, if tests are in the same library as the main library dune is able to recompile only the tests which changed.

I expected that the `libraries` stanza in `inline_tests` can be used for that but it seems like it's just for libraries for the runner (?).

**Comment by rgrinberg** [comment]
> If I understand correctly, if tests are in the same library as the main library dune is able to recompile only the tests which changed.

This is not the case. Dune only avoids rerunning inline tests in one scenario: the inline test binary remained the same. Given that this binary depends on the library itself, almost any change in the library is going to cause the binary to be re-run.

So what you are doing is the recommended practice and you aren't losing anything on recompilation.

**Reply by wokalski** [reply]
How about recompilation? I.e. let's say I have three modules in one lib called `lib1` - `A`, `B`, and `C`. The module `C` depends on `B`, `A` is independent. My understanding was that:
1. `A` changes -> only `A` is recompiled
2. `B` changes -> `B` and `C` are recompiled
3. `C` changes -> only `C` is recompiled

Now, let's assume that I also have module `D` which is in a separate library called `lib2`. `lib2` depends on `lib1`. My understanding was that if I change anything `lib1` everything in `lib2` will recompile. Is that right?

**Reply by rgrinberg** [reply]
Your understanding is correct, but it can made more accurate in dune's development mode. When a change does not modify the interface of one of these modules, no recompilation is necessary.

Sorry for the late reply.
